metadata/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+        bytes.extend(get_type_hash(
+            registry,
+            signed_extension.ty.id(),
+            &mut visited_ids,
+        ));
+        bytes.extend(get_type_hash(
+            registry,
+            signed_extension.additional_signed.id(),
+            &mut visited_ids,
+        ));
+    }
+
+    hash(&bytes)
+}
+
+/// Obtain the hash representation of a `frame_metadata::PalletMetadata`.
+pub fn get_pallet_hash(
+    registry: &PortableRegistry,
+    pallet: &frame_metadata::PalletMetadata<PortableForm>,
+    cache: &mut MetadataHasherCache,
+) -> [u8; 32] {
+    let mut bytes = vec![MetadataHashableIDs::Pallet as u8];
+    let mut visited_ids = HashSet::<u32>::new();
+
+    if let Some(cached) = cache.pallets.get(&pallet.name) {
+        return *cached
+    }
+
+    if let Some(ref calls) = pallet.calls {
+        bytes.extend(get_type_hash(registry, calls.ty.id(), &mut visited_ids));
+    }
+    if let Some(ref event) = pallet.event {
+        bytes.extend(get_type_hash(registry, event.ty.id(), &mut visited_ids));
+    }
+    for constant in pallet.constants.iter() {
+        bytes.extend(constant.name.as_bytes());
+        bytes.extend(&constant.value);
+        bytes.extend(get_type_hash(registry, constant.ty.id(), &mut visited_ids));
+    }
+    if let Some(ref error) = pallet.error {
+        bytes.extend(get_type_hash(registry, error.ty.id(), &mut visited_ids));
+    }
+    if let Some(ref storage) = pallet.storage {
+        bytes.extend(storage.prefix.as_bytes());
+        for entry in storage.entries.iter() {
+            bytes.extend(entry.name.as_bytes());
+            entry.modifier.encode_to(&mut bytes);
+            match &entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    bytes.extend(get_type_hash(registry, ty.id(), &mut visited_ids));
+                }
+                StorageEntryType::Map {
+                    hashers,
+                    key,
+                    value,
+                } => {
+                    hashers.encode_to(&mut bytes);
+                    bytes.extend(get_type_hash(registry, key.id(), &mut visited_ids));
+                    bytes.extend(get_type_hash(registry, value.id(), &mut visited_ids));
+                }
+            }
+            bytes.extend(&entry.default);
+        }
+    }
+
+    let pallet_hash = hash(&bytes);
+    cache.pallets.insert(pallet.name.clone(), pallet_hash);
+    pallet_hash
+}
+
+/// Obtain the hash representation of a `frame_metadata::RuntimeMetadataLastVersion`.
+pub fn get_metadata_hash(
+    metadata: &RuntimeMetadataLastVersion,
+    cache: &mut MetadataHasherCache,
+) -> [u8; 32] {
+    // Collect all pairs of (pallet name, pallet hash).
+    let mut pallets: Vec<(String, [u8; 32])> = metadata
+        .pallets
+        .iter()
+        .map(|pallet| {
+            let name = pallet.name.clone();
+            let hash = get_pallet_hash(&metadata.types, pallet, cache);
```

Comment:
Aah, so a reason you may have a per-pallet cache still is to re-use it here (or vice versa).

Instead of the cache, perhaps instead it's worth noting that this `get_metadata_hash` function calculates the hashes for all of the pallets as well, and since these pallet hashes may be useful, they could be returned from here too?

so for example, if `get_pallet_hash` returns eg:

```rust
pub struct MetadataHashDetails {
    pub metadata_hash: [u8; 32],
    pub pallet_hashes: HashMap<String, [u8; 32]>
}
```

then, if a subxt user wants to do the full metadata validation, we internally get this back and set some cached hash value on each pallet accordingly. Then, when the user wants to do per-pallet validation, we see that we've worked it out and use that cached value and avoid ever calling `get_pallet_hash`. If they haven't got a cached value then we call `get_pallet_hash` as needed to get the hash and populate it.

I basically think that this metadata stuff doesn't need to have any opinion on how the caching of things like that works; it can leave that up to the caller to decide, which keeps this code a little simpelr too :)

(an internal per-pallet type ID cache that the caller never sees or knows about is a good idea though I expect)